specs/codecs/wac/spec.md

```diff
@@ -0,0 +1,192 @@
+---
+title: "WAC Specification"
+navTitle: "Spec"
+---
+
+# Specification: WAC
+
+**Status: Prescriptive - Exploratory **
+
+* [Format](#format)
+* [Implementations](#implementations)
+    * [Go](#go)
+    * [Rust/WASM](#rust/wasm)
+
+WAC supports bidirectional transport to and from the [IPLD Data Model].
+In that way it is both a "complete" IPLD Data Model representation and "fitted" to the IPLD Data Model.
+Terminology taken from [Codecs and Completeness].
+
+It takes some inspiration from [Simple DAG], but differs in ways designed to make it reliably round-trip through the IPLD Data Model without loss of fidelity.
+
+The initial motivating use case for WAC is to be a codec that can reliably transport IPLD Data Model data across WebAssembly boundaries while also not being particularly complex to implement.
+
+## Format
+
+### Spec
+
+This format is a series of typing tokens, constant tokens, and inline value data.
+
+Typing tokens are proceeded with the value data for that type.
+
+Every type value can be parsed knowing only the type and without any outside context
+like the container or positional delimiters.
+
+#### Tokens
+
+TODO: Make constants in code match spec
+
+Note: The existing implementations of WAC in https://github.com/aschmahmann/wasm-ipld (Go + Rust/WASM) have different token numbers that we'll need to synchronize before merge.
+
+The tokens for the codec roughly correspond with the [IPLD Data Model Kinds]
+
+They are:
+	Null   = 0
+	True   = 1
+	False  = 2
+	Int    = 3
+	NInt   = 4
+	Float  = 5
+	String = 6
+	Bytes  = 7
+	List   = 8
+	Map    = 9
+	Link   = 10
+
+| Int | Token |
+|---|---|
+| 0 | TYPE_LINK |
+| 1 | TYPE_INTEGER |
+| 2 | TYPE_NEGATIVE_INTEGER |
+| 3 | TYPE_FLOAT |
+| 4 | TYPE_NEGATIVE_FLOAT |
+| 5 | TYPE_STRING |
+| 6 | TYPE_BINARY |
+| 7 | TYPE_MAP |
+| 8 | TYPE_LIST |
+| 9 | VALUE_NULL |
+| 10 | VALUE_TRUE |
+| 11 | VALUE_FALSE |
+
+#### TYPE_LINK
+
+```
+| 0 | CID |
+```
+
+Note: Simple-DAG put a VARINT_LENGTH before the CID indicating how long it was.
+
+CIDs are self-delimiting so this didn't seem necessary.
+
+#### TYPE_INTEGER
+
+```
+| 1 | VARINT |
```

Comment:
Probably all the varints here should be https://en.wikipedia.org/wiki/LEB128#Unsigned_LEB128 unless someone has a better idea.

While all the varint lengths will likely be capped by implementations to reasonable sizes (e.g. no one is expecting 100EiBs of data in a single WAC encoded block 🙃), I don't see why positive and negate integers shouldn't be allowed to be enormous and still called integers.

Reply:
probably worth referencing https://github.com/multiformats/unsigned-varint as the definition for it somewhere in the doc since we like both unsigned and _minimally encoded_

Reply:
Makes sense. Although I don't think we want this part of that spec `Implementations MUST restrict the size of the varint to a max of 9 bytes (63 bits).`. If someone wants a super big integer here I don't see why not, a big integer is still an integer.

Reply:
That's fine, but you should probably have language in here about the practical limits - something like: "Implementations that write integers using more than 9 bytes should not expect data to be readable by other implementations due to the commonality of 64-bit integer representations in IPLD implementations. It may be reasonable to implement a hard limit on varint representations by default but this decision is left to the implementer."